### PR TITLE
WT-5672 Rollback to stable support with start/stop aggregated durable timestamps

### PR DESCRIPTION
**Pull request by kommiharibabu**
The rollback to stable operation depends on the aggregated
timestamps that are stored as part of the checkpoint and in the
internal pages of the tree. This change adjusts/corrects the
aggregation of both start/stop durable timestamps.

Along with aggregation support fixes, the rollback to stable
is modified to check for durable start/stop timestamp while
deciding the visibility of the value.

### CODE REVIEW

**Comment by kommiharibabu** [comment]
evergreen retry

File: src/reconcile/rec_write.c [hunk 1]
```diff
@@ -2326,8 +2328,8 @@ __rec_hs_wrapup(WT_SESSION_IMPL *session, WT_RECONCILE *r)
  */
 int
 __wt_rec_cell_build_ovfl(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REC_KV *kv, uint8_t type,
-  wt_timestamp_t start_ts, uint64_t start_txn, wt_timestamp_t stop_ts, uint64_t stop_txn,
-  uint64_t rle)
+  wt_timestamp_t start_durbale_ts, wt_timestamp_t start_ts, uint64_t start_txn,
+  wt_timestamp_t stop_durbale_ts, wt_timestamp_t stop_ts, uint64_t stop_txn, uint64_t rle)
```
Comment (ddanderson) [inline src/reconcile/rec_write.c]: `durable` misspelled in these variable names.

Comment (kommiharibabu) [reply src/reconcile/rec_write.c]: Fixed.

**Comment by kommiharibabu** [comment]
@ddanderson All the existing rollback to stable tests are passed with these changes. Earlier `newest_durable_ts` is used for deciding the rollback to stable. With additional start and stop aggregated durable timestamps, the tests are passed.

**Comment by ddanderson** [comment]
@kommiharibabu, thanks, that makes sense.

**Comment by quchenhao** [comment]
This change also fixes WT-5876 but there is an issue. I fixed it in aac418a.

File: src/include/reconcile.i [hunk 2]
```diff
@@ -43,25 +43,27 @@ __wt_rec_need_split(WT_RECONCILE *r, size_t len)
...
-    *newest_durable_ts = WT_TS_NONE;
+    *start_durable_ts = WT_TS_NONE;
     *oldest_start_tsp = WT_TS_MAX;
     *oldest_start_txnp = WT_TXN_MAX;
+    *stop_durable_ts = WT_TS_NONE;
     *newest_stop_tsp = WT_TS_NONE;
     *newest_stop_txnp = WT_TXN_NONE;
     if (r->page->type == WT_PAGE_COL_FIX) {
-        *newest_durable_ts = WT_TS_NONE;
+        *start_durable_ts = WT_TS_NONE;
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/include/reconcile.i]: Fixed.

File: src/include/reconcile.i [hunk 3]
```diff
@@ -43,25 +43,27 @@ __wt_rec_need_split(WT_RECONCILE *r, size_t len)
...
         *oldest_start_tsp = WT_TS_NONE;
         *oldest_start_txnp = WT_TXN_NONE;
+        *stop_durable_ts = WT_TS_NONE;
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/include/reconcile.i]: Fixed.

**Comment by agorrod** [comment]
evergreen retry

**Comment by kommiharibabu** [comment]
Evergreen retry

**Comment by kommiharibabu** [comment]
evergreen retry

File: src/include/cell.i [hunk 4]
```diff
@@ -12,17 +12,17 @@
...
-    if (durable_start_ts > durable_stop_ts) {
-        __wt_errx(session, "a durable start timestamp %s newer than its durable stop timestamp %s",
-          __wt_timestamp_to_string(durable_start_ts, ts_string[0]),
-          __wt_timestamp_to_string(durable_stop_ts, ts_string[1]));
-        WT_ASSERT(session, durable_start_ts <= durable_stop_ts);
+    if (start_ts > durable_start_ts) {
+        __wt_errx(session, "a start timestamp %s newer than its durable start timestamp %s",
+          __wt_timestamp_to_string(start_ts, ts_string[0]),
+          __wt_timestamp_to_string(durable_start_ts, ts_string[1]));
+        WT_ASSERT(session, start_ts <= durable_start_ts);
```
Comment (keithbostic) [inline src/include/cell.i]: This could just be `WT_ERR_ASSERT`.

File: src/include/cell.i [hunk 5]
```diff
@@ -43,6 +43,15 @@ __cell_check_value_validity(WT_SESSION_IMPL *session, wt_timestamp_t durable_sta
...
     }
+
+    if (stop_ts != WT_TS_MAX && stop_ts > durable_stop_ts) {
+        __wt_errx(session,
+          "a stop timestamp %s newer than its durable "
+          "stop timestamp %s",
+          __wt_timestamp_to_string(stop_ts, ts_string[0]),
+          __wt_timestamp_to_string(durable_stop_ts, ts_string[1]));
+        WT_ASSERT(session, stop_ts <= durable_stop_ts);
+    }
```
Comment (keithbostic) [inline src/include/cell.i]: Could be `WT_ERR_ASSERT`.

File: src/include/cell.i [hunk 6]
```diff
@@ -60,17 +69,17 @@ __cell_check_value_validity(WT_SESSION_IMPL *session, wt_timestamp_t durable_sta
...
 
     __cell_check_value_validity(
-      session, durable_start_ts, durable_stop_ts, start_ts, start_txn, stop_ts, stop_txn);
+      session, durable_start_ts, start_ts, start_txn, durable_stop_ts, stop_ts, stop_txn);
 
     /* Globally visible values have no associated validity window, set a flag bit and store them. */
-    if (start_ts == WT_TS_NONE && start_txn == WT_TXN_NONE && stop_ts == WT_TS_MAX &&
-      stop_txn == WT_TXN_MAX)
+    if (durable_start_ts == WT_TS_NONE && start_ts == WT_TS_NONE && start_txn == WT_TXN_NONE &&
+      durable_stop_ts == WT_TS_NONE && stop_ts == WT_TS_MAX && stop_txn == WT_TXN_MAX)
```
Comment (keithbostic) [inline src/include/cell.i]: Why isn't `durable_stop_ts` checked against `WT_TS_MAX`?

There are a bunch of places where this happens (either testing or initialization as a default), so I'm sure there's a reason.

File: src/include/cell.i [hunk 7]
```diff
@@ -146,10 +155,28 @@ __wt_check_addr_validity(WT_SESSION_IMPL *session, wt_timestamp_t oldest_start_t
...
+        WT_ASSERT(session, oldest_start_ts <= start_durable_ts);
+    }
+    if (newest_stop_ts != WT_TS_MAX && newest_stop_ts > stop_durable_ts) {
+        __wt_errx(session,
+          "a newest stop timestamp %s newer than its durable "
+          "stop timestamp %s",
+          __wt_timestamp_to_string(newest_stop_ts, ts_string[0]),
+          __wt_timestamp_to_string(stop_durable_ts, ts_string[1]));
+        WT_ASSERT(session, newest_stop_ts <= stop_durable_ts);
+    }
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/include/cell.i]: Also, `WT_ERR_ASSERT`.

File: src/include/reconcile.i [hunk 8]
```diff
@@ -43,23 +43,23 @@ __wt_rec_need_split(WT_RECONCILE *r, size_t len)
...
      * If the page is not fixed-length column-store, where we don't maintain timestamps at all, set
      * the oldest/newest timestamps to values at the end of their expected range so they're
      * corrected as we process key/value items. Otherwise, set the oldest/newest timestamps to
      * simple durability.
      */
-    *newest_durable_ts = WT_TS_NONE;
+    *start_durable_ts = WT_TS_NONE;
     *oldest_start_tsp = WT_TS_MAX;
     *oldest_start_txnp = WT_TXN_MAX;
+    *stop_durable_ts = WT_TS_NONE;
```
Comment (keithbostic) [inline src/include/reconcile.i]: Here's the opposite, where `start_durable_ts` has a "max" of `WT_TS_NONE`, which isn't what I'd expect.

File: src/reconcile/rec_col.c [hunk 9]
```diff
@@ -632,23 +637,12 @@ __wt_rec_col_var(
...
-    if ((addr = pageref->addr) == NULL)
-        newest_durable_ts = WT_TS_NONE;
-    else if (__wt_off_page(pageref->home, addr))
-        newest_durable_ts = addr->stop_durable_ts;
-    else {
-        __wt_cell_unpack(session, pageref->home, pageref->addr, vpack);
-        newest_durable_ts = vpack->newest_stop_durable_ts;
-    }
-
     /* Set the "last" values to cause failure if they're not set. */
```
Comment (keithbostic) [inline src/reconcile/rec_col.c]: I don't understand why this code is no longer needed.

Similar code was removed in row-store reconciliation.

File: src/include/reconcile.i (hunk 8 above)
Comment (kommiharibabu) [reply src/include/reconcile.i]: As per the discussion in the ticket WT-5668, as both start and stop durable timestamps will hold the maximum value of timestamp on the page, so it is better to use the same initial value for both of them for easier understanding. And we followed the same notation irrespective of the value or address cell.

File: src/include/cell.i (hunk 6 above)
Comment (kommiharibabu) [reply src/include/cell.i]: we want to follow the same initial value for all the durable timestamps across the code as both of them will store the maximum value of the start/stop timestamps. Refer WT-5668 for more details.

File: src/reconcile/rec_col.c (hunk 9 above)
Comment (kommiharibabu) [reply src/reconcile/rec_col.c]: Earlier we didn't store the durable timestamp as part of the `value` cell and only it is saved in the `address` cell. Whenever the same page is getting reconciled again, we take the `newest durable timestamp` stored in the address cell as the default value for each `value` cell on the page. But now as we are storing the durable timestamp as part of the `value` cell, we don't need this special code.

File: src/include/cell.i (hunk 4 above)
Comment (kommiharibabu) [reply src/include/cell.i]: Changed.

File: src/include/cell.i (hunk 5 above)
Comment (kommiharibabu) [reply src/include/cell.i]: Changed

File: src/include/cell.i [hunk 10]
```diff
@@ -43,6 +43,15 @@ __cell_check_value_validity(WT_SESSION_IMPL *session, wt_timestamp_t durable_sta
           start_txn, stop_txn);
         WT_ASSERT(session, start_txn <= stop_txn);
     }
+
+    if (stop_ts != WT_TS_MAX && stop_ts > durable_stop_ts) {
+        __wt_errx(session,
+          "a stop timestamp %s newer than its durable "
+          "stop timestamp %s",
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/include/cell.i]: Corrected most of the splits.

File: src/include/cell.i (hunk 7 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/include/cell.i]: Changed to use WT_ERR_ASSERT.

**Pull request merged**